In [51]:
import pandas as pd
import seaborn as sns
import numpy as np
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
from sklearn.metrics import mean_squared_error

casualty =  pd.read_csv("/Users/swimmingcircle/cs156_code/session03/casualty_train.csv")

## Classification Model 

In [52]:
casualty.head()

casualty_class  gender age  severe pedestrian_location pedestrian_movement  \
0      passenger  female  33   False                 NaN                 NaN   
1      passenger  female  20   False                 NaN                 NaN   
2      passenger    male  52   False                 NaN                 NaN   
3      passenger  female  17   False                 NaN                 NaN   
4      passenger  female  20   False                 NaN                 NaN   

       travel  year  
0   motorbike  2007  
1         car  2005  
2         car  2006  
3  pedestrian  2012  
4   motorbike  2010

In [53]:
print('type of casualty class:', casualty['casualty_class'].unique())
print('type of gender:', casualty['gender'].unique())
print('type of gender:', casualty['age'].unique())
print('type of severe:', casualty['severe'].unique())
print('type of pedestrian_location:', casualty['pedestrian_location'].unique())
print('type of pedestrian_movement:', casualty['pedestrian_movement'].unique())
print('type of travel:', casualty['travel'].unique())

type of casualty class: ['passenger' 'pedestrian' 'driver']
type of gender: ['female' 'male']
type of gender: ['33' '20' '52' '17' '40' '23' '45' '8' '7' 'Unknown' '43' '34' '32' '31'
 '29' '50' '4' '12' '25' '27' '38' '39' '51' '37' '86' '10' '46' '36' '28'
 '60' '57' '65' '21' '68' '3' '42' '18' '26' '24' '44' '61' '48' '5' '11'
 '35' '58' '13' '30' '56' '22' '84' '1' '15' '49' '63' '53' '41' '74' '73'
 '77' '19' '71' '75' '16' '47' '14' '64' '76' '2' '81' '79' '54' '67' '9'
 '83' '66' '87' '78' '80' '69' '59' '93' '70' '72' '88' '6' '85' '55' '82'
 '62' '89' '92' '0' '90' '98' '95' '91' '96' '97' '94' '99']
type of severe: [False  True]
type of pedestrian_location: [nan 'In road' 'Close to Crossing' 'Not on Crossing' 'Unknown'
 'Pedestrian Crossing' 'Footpath' 'Zig-Zag']
type of pedestrian_movement: [nan 'In Rd Not Crossing' 'Drivers N/Side' 'Unknown Or Other'
 'Drivers N/Side Msk' 'Drivers O/Side' 'Back To Traffic'
 'Drivers O/Side Msk' 'Facing Traffic' 'In Rd Not Crossing Msk']
ty

In [54]:
casualty = casualty[casualty.age != 'Unknown']

In [55]:
# casualty.dropna(inplace=True)


In [56]:
y = casualty['severe']
X = casualty.loc[:,  casualty.columns!= 'severe']


In [57]:
def label_encoding(df, cols):
    le = LabelEncoder()
    for col in cols:
        df.loc[:,col] = le.fit_transform(df.loc[:,col])
    return df
cols = ['casualty_class', 'gender','pedestrian_location',
       'pedestrian_movement', 'travel', 'year']
X = label_encoding(X, cols)
X

/Users/swimmingcircle/Library/Python/3.9/lib/python/site-packages/pandas/core/indexing.py:1773: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(ilocs[0], value, pi)


casualty_class  gender age  pedestrian_location  pedestrian_movement  \
0                    1       0  33                    7                    9   
1                    1       0  20                    7                    9   
2                    1       1  52                    7                    9   
3                    1       0  17                    7                    9   
4                    1       0  20                    7                    9   
...                ...     ...  ..                  ...                  ...   
232830               1       0  40                    7                    9   
232831               1       0  41                    7                    9   
232832               2       0  45                    0                    1   
232833               2       1  10                    0                    2   
232834               1       0  37                    7                    9   

        travel  year  
0            3     2  
1            2     0  
2            2     1  
3            5     7  
4            3     5  
...        ...   ...  
232830       2     9  
232831       5     8  
232832       0     5  
232833       0     4  
232834       2     9  

[216610 rows x 7 columns]

In [58]:
X.astype(int)

casualty_class  gender  age  pedestrian_location  pedestrian_movement  \
0                    1       0   33                    7                    9   
1                    1       0   20                    7                    9   
2                    1       1   52                    7                    9   
3                    1       0   17                    7                    9   
4                    1       0   20                    7                    9   
...                ...     ...  ...                  ...                  ...   
232830               1       0   40                    7                    9   
232831               1       0   41                    7                    9   
232832               2       0   45                    0                    1   
232833               2       1   10                    0                    2   
232834               1       0   37                    7                    9   

        travel  year  
0            3     2  
1            2     0  
2            2     1  
3            5     7  
4            3     5  
...        ...   ...  
232830       2     9  
232831       5     8  
232832       0     5  
232833       0     4  
232834       2     9  

[216610 rows x 7 columns]

In [59]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 0)

### Choice of classification model: Logistic vs KNN 

- Logistic regression: parametric model 
- KNN:  non parametric model 

If we need examine the parameters e.g. have interpretable meaning for ceofs, run some tests to obtain p-value and discard non significant paramters. We will need to use parametric. 

Since we need to interpret the parameters, we need to choose parametric model. I choose logistic regression. 

In [60]:
from sklearn.linear_model import LogisticRegression
log_model = LogisticRegression(random_state=0)
log_model.fit(X_train, y_train)
y_pred = log_model.predict(X_test)

In [61]:
coeff_parameter = pd.DataFrame(log_model.coef_.reshape(-1,1), X.columns,columns=['Coefficient'])
coeff_parameter.loc['intercept', :] = log_model.intercept_
coeff_parameter

Coefficient
casualty_class          0.260269
gender                 -0.273917
age                     0.007460
pedestrian_location    -0.122312
pedestrian_movement    -0.074118
travel                  0.129585
year                   -0.072317
intercept              -1.257919

### Model Evaluation

#### Confusion matrix

In [76]:
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test, y_pred)

array([[38608,     0],
       [ 4714,     0]])

We see that true negative = 38608, and false negative = 4714. This showcases we classify every things as negative. 

In [85]:
from sklearn.metrics import accuracy_score
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
print('accuracy score is', accuracy_score(y_test, y_pred))
print('precision score is', precision_score(y_test, y_pred, average=None))
print('recall score is', recall_score(y_test, y_pred, average=None))

accuracy score is 0.891186925811366
precision score is [0.89118693 0.        ]
recall score is [1. 0.]


/Users/swimmingcircle/Library/Python/3.9/lib/python/site-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [83]:
from sklearn import metrics
print(metrics.classification_report(y_test, y_pred))

              precision    recall  f1-score   support

       False       0.89      1.00      0.94     38608
        True       0.00      0.00      0.00      4714

    accuracy                           0.89     43322
   macro avg       0.45      0.50      0.47     43322
weighted avg       0.79      0.89      0.84     43322



/Users/swimmingcircle/Library/Python/3.9/lib/python/site-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/Users/swimmingcircle/Library/Python/3.9/lib/python/site-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/Users/swimmingcircle/Library/Python/3.9/lib/python/site-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start,

### Interpret the classification report table

Precision = Accuracy of positive predictions = $\frac{TP}{(TP + FP)}$

Recall = Fraction of positives that were correctly identified.
Recall = $\frac{TP}{(TP + FN)}$

F1 Score = $\frac{2*(Recall * Precision)}{(Recall + Precision)}$


### Question
- What's the difference between precision, recall, and accuracy f1-score? 
- How to interpret the parameter in logistic regression 

In [ ]:
casualty_test =  pd.read_csv("/Users/swimmingcircle/cs156_code/session03/casualty_test.csv")

In [ ]:
casualty_test

casualty_class  gender age  severe pedestrian_location  \
0          passenger    male  32    True                 NaN   
1          passenger    male  34   False                 NaN   
2          passenger  female  57   False                 NaN   
3          passenger    male  37   False                 NaN   
4          passenger  female  23   False                 NaN   
...              ...     ...  ..     ...                 ...   
58203      passenger  female  45    True                 NaN   
58204      passenger  female  38    True                 NaN   
58205      passenger    male  23    True                 NaN   
58206         driver  female  23   False                 NaN   
58207      passenger    male  42   False                 NaN   

      pedestrian_movement      travel  year  
0                     NaN  pedestrian  2013  
1                     NaN  pedestrian  2008  
2                     NaN         car  2010  
3                     NaN         car  2011  
4                     NaN         car  2007  
...                   ...         ...   ...  
58203                 NaN   motorbike  2010  
58204                 NaN   motorbike  2014  
58205                 NaN  pedestrian  2011  
58206                 NaN         car  2010  
58207                 NaN         car  2005  

[58208 rows x 8 columns]

In [ ]:
print('type of casualty class:', casualty_test['casualty_class'].unique())
print('type of gender:', casualty_test['gender'].unique())
print('type of gender:', casualty_test['age'].unique())
print('type of severe:', casualty_test['severe'].unique())
print('type of pedestrian_location:', casualty_test['pedestrian_location'].unique())
print('type of pedestrian_movement:', casualty_test['pedestrian_movement'].unique())
print('type of travel:', casualty_test['travel'].unique())

type of casualty class: ['passenger' 'pedestrian' 'driver']
type of gender: ['male' 'female']
type of gender: ['32' '34' '57' '37' '23' '20' '44' '42' '39' '24' '46' '65' '43' '31'
 '33' 'Unknown' '88' '26' '22' '13' '54' '29' '35' '38' '30' '45' '40'
 '28' '27' '59' '19' '21' '25' '14' '41' '58' '6' '87' '17' '50' '18' '68'
 '56' '11' '64' '67' '47' '55' '7' '10' '16' '36' '63' '9' '76' '2' '53'
 '3' '5' '73' '83' '60' '49' '15' '72' '51' '8' '78' '80' '84' '75' '66'
 '12' '61' '52' '82' '70' '48' '79' '86' '81' '1' '74' '69' '77' '85' '62'
 '4' '71' '0' '92' '90' '89' '95' '96' '94' '91' '93' '97' '99']
type of severe: [ True False]
type of pedestrian_location: [nan 'Not on Crossing' 'In road' 'Unknown' 'Footpath'
 'Pedestrian Crossing' 'Close to Crossing' 'Zig-Zag']
type of pedestrian_movement: [nan 'Drivers N/Side Msk' 'Unknown Or Other' 'Drivers N/Side'
 'Drivers O/Side' 'Drivers O/Side Msk' 'In Rd Not Crossing'
 'In Rd Not Crossing Msk' 'Back To Traffic' 'Facing Traffic']
type of

In [ ]:
casualty_test = casualty_test[casualty_test.age != 'Unknown']
y = casualty_test['severe']
X = casualty_test.loc[:,  casualty_test.columns!= 'severe'].values
le = LabelEncoder()




X[:,0] = le.fit_transform(X[:,0])
X[:,1] = le.fit_transform(X[:,1])
X[:,3] = le.fit_transform(X[:,3])
X[:,4] = le.fit_transform(X[:,4])
X[:,5] = le.fit_transform(X[:,5])
X[:,6] = le.fit_transform(X[:,6])
X.astype(int)

array([[ 1,  1, 32, ...,  9,  5,  8],
       [ 1,  1, 34, ...,  9,  5,  3],
       [ 1,  0, 57, ...,  9,  2,  5],
       ...,
       [ 1,  1, 23, ...,  9,  5,  6],
       [ 0,  0, 23, ...,  9,  2,  5],
       [ 1,  1, 42, ...,  9,  2,  0]])

In [ ]:
log_model.fit(X, y)
y_pred = log_model.predict(X)
print(metrics.classification_report(y, y_pred))


              precision    recall  f1-score   support

       False       0.89      1.00      0.94     48369
        True       0.00      0.00      0.00      5908

    accuracy                           0.89     54277
   macro avg       0.45      0.50      0.47     54277
weighted avg       0.79      0.89      0.84     54277



/Users/swimmingcircle/Library/Python/3.9/lib/python/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/Users/swimmingcircle/Library/Python/3.9/lib/python/site-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/Users/swimmingcircle/Library/Python/3.9/lib/python/site-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWar

In [ ]:
print(casualty['severe'].value_counts())
print(casualty_test['severe'].value_counts())

False    193099
True      23511
Name: severe, dtype: int64
False    48369
True      5908
Name: severe, dtype: int64


For both training and testing, more False than True exists. We have significant 